### ISSUE TITLE
Race condition in CatalogSafetyInitializer

### ISSUE THREAD

**Issue by marksimu** [killbill/killbill#696]
We have seen a few random occurrences of this error.  They've begun with 0.18.2:

```
org.killbill.notificationq.NotificationQueueException: java.lang.RuntimeException: Failed during catalog initialization :
        at org.killbill.notificationq.NotificationQueueDispatcher.handleNotificationWithMetrics(NotificationQueueDispatcher.java:205)
        at org.killbill.notificationq.dispatching.NotificationCallableCallback.dispatch(NotificationCallableCallback.java:51)
        at org.killbill.notificationq.dispatching.NotificationCallableCallback.dispatch(NotificationCallableCallback.java:32)
        at org.killbill.queue.dispatching.Dispatcher$CallableQueue.call(Dispatcher.java:103)
        at org.killbill.queue.dispatching.Dispatcher$CallableQueue.call(Dispatcher.java:86)
        at org.killbill.commons.concurrent.WrappedCallable.call(WrappedCallable.java:42)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at org.killbill.commons.concurrent.WrappedRunnable.run(WrappedRunnable.java:40)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: Failed during catalog initialization :
        at org.killbill.billing.catalog.CatalogSafetyInitializer.initializeNonRequiredNullFieldsWithDefaultValue(CatalogSafetyInitializer.java:67)
        at org.killbill.billing.catalog.DefaultProduct.initialize(DefaultProduct.java:155)
        at org.killbill.billing.catalog.StandaloneCatalog.initialize(StandaloneCatalog.java:312)
        at org.killbill.billing.catalog.plugin.StandaloneCatalogMapper.toStandaloneCatalog(StandaloneCatalogMapper.java:119)
        at org.killbill.billing.catalog.plugin.VersionedCatalogMapper.toStandaloneCatalogWithPriceOverride(VersionedCatalogMapper.java:63)
        at org.killbill.billing.catalog.plugin.VersionedCatalogMapper.toVersionedCatalog(VersionedCatalogMapper.java:56)
        at org.killbill.billing.catalog.caching.EhCacheCatalogCache.getCatalogFromPlugins(EhCacheCatalogCache.java:139)
        at org.killbill.billing.catalog.caching.EhCacheCatalogCache.getCatalog(EhCacheCatalogCache.java:95)
        at org.killbill.billing.catalog.DefaultCatalogService.getCatalog(DefaultCatalogService.java:104)
        at org.killbill.billing.catalog.DefaultCatalogService.getFullCatalog(DefaultCatalogService.java:95)
        at org.killbill.billing.subscription.engine.dao.DefaultSubscriptionDao.createSubscriptionForInternalUse(DefaultSubscriptionDao.java:976)
        at org.killbill.billing.subscription.engine.dao.DefaultSubscriptionDao.buildBundleSubscriptions(DefaultSubscriptionDao.java:813)
        at org.killbill.billing.subscription.engine.dao.DefaultSubscriptionDao.buildSubscription(DefaultSubscriptionDao.java:785)
        at org.killbill.billing.subscription.engine.dao.DefaultSubscriptionDao.getSubscriptionFromId(DefaultSubscriptionDao.java:313)
        at org.killbill.billing.subscription.api.svcs.DefaultSubscriptionInternalApi.getSubscriptionFromId(DefaultSubscriptionInternalApi.java:482)
        at org.killbill.billing.invoice.notification.DefaultNextBillingDateNotifier$1.handleReadyNotification(DefaultNextBillingDateNotifier.java:80)
        at org.killbill.notificationq.NotificationQueueDispatcher.handleNotificationWithMetrics(NotificationQueueDispatcher.java:203)
        ... 10 common frames omitted
Caused by: java.lang.IllegalAccessException: Class org.killbill.billing.catalog.CatalogSafetyInitializer can not access a member of class org.killbill.billing.catalog.DefaultProduct with modifiers "private"
        at sun.reflect.Reflection.ensureMemberAccess(Reflection.java:102)
        at java.lang.reflect.AccessibleObject.slowCheckMemberAccess(AccessibleObject.java:296)
        at java.lang.reflect.AccessibleObject.checkAccess(AccessibleObject.java:288)
        at java.lang.reflect.Field.get(Field.java:390)
        at org.killbill.billing.catalog.CatalogSafetyInitializer.initializeArrayIfNull(CatalogSafetyInitializer.java:116)
        at org.killbill.billing.catalog.CatalogSafetyInitializer.initializeNonRequiredNullFieldsWithDefaultValue(CatalogSafetyInitializer.java:53)
        ... 26 common frames omitted
```

I believe this is due to https://github.com/killbill/killbill/commit/bcab8394aecede394fb15e9de5e083ad1c89acec.  With this commit the Field objects are now cached and shared by all threads.

```
f.setAccessible(true);
if (f.get(obj) == null) {
	f.set(obj, getZeroLengthArrayInitializer(f));
}
f.setAccessible(false);
```

The exception occurs when one thread executes f.setAccessible(false) while another thread is executing f.get(obj).

**Comment by sbrossie** [comment]
One easy way to fix is to make those `initializeFieldWithValue` and `initializeArrayIfNull` methods`synchronized`. It might have a small perf impact when multiple threads load catalog at the same time (which is your case here).

On that topic, i which we had pursued the caching discussion we started, for that catalog plugin api because although we have improved perf quite a bit, this is quite barbaric to have to recreate this large catalog every single time there is a component requiring the catalog.
